test/execution/atomics_test.cpp

```diff
@@ -0,0 +1,513 @@
+   *=========================
+   */
+
+  std::default_random_engine generator;
+  const uint32_t num_iters = 100;
+  const uint32_t num_threads = 16;  // Number of bits in test
+  common::WorkerPool thread_pool(num_threads, {});
+
+  for (uint32_t iter = 0; iter < num_iters; ++iter) {
+    std::atomic<uint16_t> target = 0;
+    auto workload = [&](uint32_t thread_id) {
+      uint16_t expected;
+      bool success = false;
+      do {
+        expected = thread_id;
+        EXPECT_FALSE(success);
+        success = cmpxchg(reinterpret_cast<uint16_t *>(&target), &expected, static_cast<uint16_t>(thread_id + 1));
+        EXPECT_LE(expected, thread_id);  // Out-of-order exchange occurred
+        EXPECT_TRUE(!success || expected == thread_id);
+      } while (expected != thread_id);
+      EXPECT_TRUE(success);
+    };
+
+    MultiThreadTestUtil::RunThreadsUntilFinish(&thread_pool, num_threads, workload);
+    EXPECT_EQ(target.load(), num_threads);
+  }
+}
+
+// NOLINTNEXTLINE
+TEST_F(AtomicsTest, AtomicCompareExchange4) {
+  // Setup the compilation environment
+  sema::ErrorReporter error_reporter(&region_);
+  ast::AstNodeFactory factory(&region_);
+  ast::Context context(&region_, &error_reporter);
+
+  auto src = std::string(
+      "fun cmpxchg(dest: *uint32, expected: *uint32, desired: uint32) -> bool {\n"
+      "  var x = @atomicCompareExchange(dest, expected, desired)\n"
+      "  return x\n"
+      "}");
+
+  // Compile it...
+  auto input = compiler::Compiler::Input("Atomic Definitions", &context, &src);
+  auto module = compiler::Compiler::RunCompilationSimple(input);
+  EXPECT_STREQ(error_reporter.SerializeErrors().c_str(), "");
+  ASSERT_FALSE(module == nullptr);
+
+  // The function should exist
+  std::function<bool(uint32_t *, uint32_t *, uint32_t)> cmpxchg;
+  EXPECT_TRUE(module->GetFunction("cmpxchg", vm::ExecutionMode::Interpret, &cmpxchg));
+
+  /*=========================
+   *= Run correctness tests =
+   *=========================
+   */
+
+  std::default_random_engine generator;
+  const uint32_t num_iters = 100;
+  const uint32_t num_threads = 32;  // Number of bits in test
+  common::WorkerPool thread_pool(num_threads, {});
+
+  for (uint32_t iter = 0; iter < num_iters; ++iter) {
+    std::atomic<uint32_t> target = 0;
+    auto workload = [&](uint32_t thread_id) {
+      uint32_t expected;
+      bool success = false;
+      do {
+        expected = thread_id;
+        EXPECT_FALSE(success);
+        success = cmpxchg(reinterpret_cast<uint32_t *>(&target), &expected, static_cast<uint32_t>(thread_id + 1));
+        EXPECT_LE(expected, thread_id);  // Out-of-order exchange occurred
+        EXPECT_TRUE(!success || expected == thread_id);
+      } while (expected != thread_id);
+      EXPECT_TRUE(success);
+    };
+
+    MultiThreadTestUtil::RunThreadsUntilFinish(&thread_pool, num_threads, workload);
+    EXPECT_EQ(target.load(), num_threads);
+  }
+}
+
+// NOLINTNEXTLINE
+TEST_F(AtomicsTest, AtomicCompareExchange8) {
```

Comment:
These tests are very very similar. With the exception of the types (both C++ and TPL) they operate on, they look identical. Can you think of a way to refactor them?

Reply:
I might be able to make the inner part of each "class" of test templated on the integer type and takes the TPL code as the argument?

Reply:
How does it look now?

Reply:
Much better. Now, rather than passing the whole source to the testing functions, can you pass just the primitive TPL type as a string and format the source based on this type using `fmt::format()`?

Reply:
Will do.